ruler/ruler.py

```diff
@@ -1,88 +1,202 @@
 
     '''
     Users sometimes need to discard certain marginal regions, the withds of which are specified by margin_size.
     The first pair of elements correspond to the first dimension, the second pair of elements correspond to the second dimension,
     and the third pair of elements, if there are, correspond to the third dimension.
     '''
     margin_size = np.reshape(margin_size,(-1,2))
+    if len(margin_size[:,0])>dims:
+        margin_size = margin_size[0:dims,:]
+        print('Warning: The row number of margin_size is larger than the dimension; The redundant row(s) will be discarded.')
     return phys_size, margin_size
 
-def minimum_length(arr,phys_size,margin_size=np.zeros((3,2)),threshold=0.5,proj_strength=10**6,len_arr=None):
+
+def minimum_length(arr,phys_size,margin_size=np.zeros((2,2)),measure_region='solid',pad_mode=None):
     '''
-    Compute the minimum length scale in a design pattern. The design pattern is arr.
-    The physical size is phys_size. The size of borders that need to be discarded is margin_size.
-    The binary values of pixels are thresholded by threshold.
-    The projection strength in some morphological operations is proj_strength.
+    Compute the minimum length scale in a design pattern. The design pattern is 'arr'.
+    The physical size is 'phys_size'. The size of borders that need to be discarded is 'margin_size'.
     '''
 
     phys_size, margin_size = ruler_initialize(arr,phys_size,margin_size)
+    shorter_entire_side = min(phys_size) # shorter side of the entire design region
+    all_pixel_sides = pixel_size(arr,phys_size)
+    shorter_pixel_side = min(all_pixel_sides) # shorter side of a pixel
     dims = len(phys_size) # dimension, should be 2 or 3
-    arr = binarize(arr,threshold)
+
+    '''
+    If all elements in the array are the same,
+    the code simply regards the shorter side of the entire pattern as the minimum length scale,
+    regardless of whether the pattern is solid or void.
+    '''
+    if len(np.unique(arr))==1:
+        return shorter_entire_side
+
+    '''
+    Design patterns with non-binary values are thresholded by 'binarize_threshold'.
+    The appearance of interior solid pixels is thresholded by 'interior_pixel_threshold'.
+    The projection strength in some morphological operations is 'proj_strength'.
+    '''
+    binarize_threshold, interior_pixel_threshold, proj_strength = 0.5, 0.5, 10**6
+    arr = binarize(arr,binarize_threshold)
     margin_number = margin(arr,phys_size,margin_size)
 
-    if np.array(len_arr).any(): # search the minimum length scale within a length array "len_arr"
-        radius_list = sorted(list(np.abs(len_arr)/2))
-        for radius in radius_list:
-            # difference between open and close
-            diff_image = abs(open_operator(arr,radius,phys_size,proj_strength)-close_operator(arr,radius,phys_size,proj_strength))
```

Comment:
I think we should continue to use the open–close algorithm when the user wants both the solid and void lengthscales — it is actually a really nice property that this is symmetric to inversion of solid and void!